File: LeetCode/931/solution-recursive-dp.cpp
// 931. Minimum Falling Path Sum
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        int row = matrix.size();
        int col = matrix[0].size();

        vector<vector<int>> dp(row, vector<int>(col, INT_MAX));

        int ans = INT_MAX;
        for (int i=0; i<col; i++){
            ans = min(ans, solve(0, i, matrix, dp));
        }

        return ans;
    }

    int solve(int row, int col, vector<vector<int>>& mat, vector<vector<int>>& dp){
        int ROW = mat.size();
        int COL = mat[0].size();

        if(row==ROW) return 0;
        if(col < 0 || col > COL-1) return INT_MAX;

        if(dp[row][col] != INT_MAX) return dp[row][col];

        int left = solve(row+1, col-1, mat, dp);
        int down =  solve(row+1, col, mat, dp);
        int right = solve(row+1, col+1, mat, dp);

        int ret = mat[row][col] + min({ left, down, right});

        return dp[row][col] = ret;
    }
};

File: LeetCode/931/solution-recursive-dp.java
// 931. Minimum Falling Path Sum
class Solution {
    private int m, n;
    private int[][] dp = new int[101][101];

    public int minFallingPathSum(int[][] matrix) {
        m = matrix.length;
        n = matrix[0].length;

        for (int[] row : dp) {
            Arrays.fill(row, -10000);
        }

        int ans = Integer.MAX_VALUE;
        for (int i=0; i<n; i++){
            ans = Math.min(ans, solve(0, i, matrix));
        }

        return ans;
    }

    int solve(int row, int col, int[][] mat){

        if(row==m) return 0;
        if(col < 0 || col > n-1) return Integer.MAX_VALUE;

        if(dp[row][col] != -10000) return dp[row][col];

        int left = solve(row+1, col-1, mat);
        int down =  solve(row+1, col, mat);
        int right = solve(row+1, col+1, mat);

        int ret = mat[row][col] + Math.min(left, Math.min(down, right));

        return dp[row][col] = ret;
    }
}